### PR TITLE
Reformat determine molecules

### PR DESCRIPTION
**Pull request by afshinfard**
Code reformatting for the molecule stage:
- lesser lines of codes
- easy to add new strategies

Tested for .k3 and bi-connected. the results were totally identical

### CODE REVIEW

File: data/Makefile [hunk 1]
```diff
@@ -772,17 +772,19 @@ min_path_size=200
 	mlr --tsvlite filter '$$n >= $n' $< >$@
 
 # Separate barcodes into molecules.
-mol_strategy=1
+mol_strategy="bc"
```
Comment (sjackman) [inline data/Makefile]:
```suggestion
mol_strategy=bc
```
No need for quotes here.

File: data/Makefile [hunk 2]
```diff
@@ -772,17 +772,19 @@ min_path_size=200
...
 %.mol.tsv: %.tsv
 	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=$(mol_strategy) $< >$@
 
 # Separate barcodes into molecules using K3-clique community detection.
+mol_strategy2="bc.k3"
```
Comment (sjackman) [inline data/Makefile]:
```suggestion
mol_strategy2=bc.k3
```

File: data/Makefile [hunk 3]
```diff
@@ -772,17 +772,19 @@ min_path_size=200
...
 %.k3.mol.tsv: %.tsv
-	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=2 $< >$@
+	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=$(mol_strategy2) $< >$@
 
 # Separate barcodes into molecules using Louvain community detection.
+mol_strategy3="bc.louvain"
```
Comment (sjackman) [inline data/Makefile]:
```suggestion
mol_strategy3=bc.louvain
```

File: physlr/physlr.py [hunk 4]
```diff
@@ -1585,31 +1562,19 @@ def determine_molecules_process(u):
...
-               "(pipeline: bi-connected + partition + bi-connected + k-cliques + merge)",
-            6: "\n\tStrategy: "
-               "Community detection by extensive community-detection"
-               "\n\t(bi-connected + k3-clique + cosine similarity (with/out squaring)."
-        }
-        if self.args.strategy not in strategy_switcher:
-            exit("\033[93m Wrong input argument: --separation-strategy!\033[0m")
+        alg_list = self.args.strategy.split(".")
+        if not alg_list:
+            exit("\033[91m Input argument not provided: --separation-strategy!\033[0m")
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
            exit("Error: physlr molecule: missing parameter --separation-strategy")
```
No need for exclamation marks in error messages. Include the word `Error` and the name of the program in the error message. I do actually like ANSI colours, but if we want to use them, we need to factor them out into a function (in a separate PR).

File: physlr/physlr.py [hunk 5]
```diff
@@ -1585,31 +1562,19 @@ def determine_molecules_process(u):
...
+        if not set(alg_list).issubset(
+                {"bc", "k3", "cos", "sqCos", "sqcos", "louvain", "distributed"}):
+            exit("\033[91m Wrong input argument: --separation-strategy!\033[0m")
```
Comment (sjackman) [inline physlr/physlr.py]: Same suggestion as above, and also print the values in `alg_list` that are not in the whitelist.

File: physlr/physlr.py [hunk 6]
```diff
@@ -1555,22 +1519,35 @@ def determine_molecules_extensive(g, u):
...
-            communities = Physlr.determine_molecules_bc_louvain(g, u)
-        elif strategy == 4:  # bi-connected + sqCos
-            communities = Physlr.determine_molecules_bc_cosine_of_squared(g, u)
-        elif strategy == 5:  # bi-connected + partition + bi-connected + k-cliques + merge
-            communities = Physlr.determine_molecules_partition_split_merge(g, u)
-        elif strategy == 6:  # extensive detection
-            communities = Physlr.determine_molecules_extensive(g, u)
-        else:
-            exit("\033[93m Wrong input argument: --separation-strategy!\033[0m")
+        alg_list = strategy.split(".")
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
        alg_list = strategy.split("+")
```
Let's use `+` rather than `.`.

File: physlr/physlr.py [hunk 7]
```diff
@@ -1585,31 +1562,19 @@ def determine_molecules_process(u):
...
-            5: "\n\tStrategy: "
-               "Fast Community detection {partition + detect + merge}\n\t"
-               "(pipeline: bi-connected + partition + bi-connected + k-cliques + merge)",
-            6: "\n\tStrategy: "
-               "Community detection by extensive community-detection"
-               "\n\t(bi-connected + k3-clique + cosine similarity (with/out squaring)."
-        }
-        if self.args.strategy not in strategy_switcher:
-            exit("\033[93m Wrong input argument: --separation-strategy!\033[0m")
+        alg_list = self.args.strategy.split(".")
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
        alg_list = self.args.strategy.split("+")
```

File: physlr/physlr.py [hunk 8]
```diff
@@ -1585,31 +1562,19 @@ def determine_molecules_process(u):
...
+            exit("\033[91m Wrong input argument: --separation-strategy!\033[0m")
+
         gin = self.read_graph(self.args.FILES)
         Physlr.filter_edges(gin, self.args.n)
         print(
             int(timeit.default_timer() - t0),
-            "Separating barcodes into molecules",
-            strategy_switcher.get(self.args.strategy),
+            "Separating barcodes into molecules using the following algorithm(s):\n\t",
+            self.args.strategy.replace(".", " + "),
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
```

File: physlr/physlr.py [hunk 9]
```diff
@@ -2154,8 +2119,10 @@ def parse_arguments():
             "-w", "--window", action="store", dest="w", type=int,
             help="number of k-mers in a window of size k + w - 1 bp")
         argparser.add_argument(
-            "--separation-strategy", action="store", dest="strategy", type=int, default=1,
-            help="strategy id, for barcode to molecule separation [1]")
+            "--separation-strategy", action="store", dest="strategy", default="bc.k3",
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
            "--separation-strategy", action="store", dest="strategy", default="bc+k3",
```

File: physlr/physlr.py [hunk 10]
```diff
@@ -2154,8 +2119,10 @@ def parse_arguments():
...
+            help="strategy for barcode to molecule separation [bc.k3.bc]. Use a combination"
+                 " of bc, k3, cos, sqCos, louvain, and distributed"
+                 " concatenated with dots (example:bc.k3.bc)")
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
                 " concatenated with plus signs")
```

File: physlr/physlr.py [hunk 11]
```diff
@@ -2154,8 +2119,10 @@ def parse_arguments():
             "-w", "--window", action="store", dest="w", type=int,
             help="number of k-mers in a window of size k + w - 1 bp")
         argparser.add_argument(
-            "--separation-strategy", action="store", dest="strategy", type=int, default=1,
-            help="strategy id, for barcode to molecule separation [1]")
+            "--separation-strategy", action="store", dest="strategy", default="bc.k3",
+            help="strategy for barcode to molecule separation [bc.k3.bc]. Use a combination"
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
            help="strategy for barcode to molecule separation [bc+k3]. Use a combination"
```

**Comment by afshinfard** [comment]
and I forgot that I can remove `determine_molecules_extensive` function and we can make it by listing the algorithms we want. So lesser lines again.

**Comment by afshinfard** [comment]
Ready for review. and I am testing for identical output...

File: physlr/physlr.py [hunk 12]
```diff
@@ -1585,31 +1527,21 @@ def determine_molecules_process(u):
...
-            5: "\n\tStrategy: "
-               "Fast Community detection {partition + detect + merge}\n\t"
-               "(pipeline: bi-connected + partition + bi-connected + k-cliques + merge)",
-            6: "\n\tStrategy: "
-               "Community detection by extensive community-detection"
-               "\n\t(bi-connected + k3-clique + cosine similarity (with/out squaring)."
-        }
-        if self.args.strategy not in strategy_switcher:
-            exit("\033[93m Wrong input argument: --separation-strategy!\033[0m")
+        alg_white_list = {"bc", "k3", "cos", "sqCos", "sqcos", "louvain", "distributed"}
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
        alg_white_list = {"bc", "k3", "cos", "sqcos", "louvain", "distributed"}
```

File: physlr/physlr.py [hunk 13]
```diff
@@ -2154,8 +2086,10 @@ def parse_arguments():
             "-w", "--window", action="store", dest="w", type=int,
             help="number of k-mers in a window of size k + w - 1 bp")
         argparser.add_argument(
-            "--separation-strategy", action="store", dest="strategy", type=int, default=1,
-            help="strategy id, for barcode to molecule separation [1]")
+            "--separation-strategy", action="store", dest="strategy", default="bc+k3",
+            help="strategy for barcode to molecule separation [bc+k3]. Use a combination"
+                 " of bc, k3, cos, sqCos, louvain, and distributed"
```
Comment (sjackman) [inline physlr/physlr.py]:
```suggestion
                 " of bc, k3, cos, sqcos, louvain, and distributed"
```

File: data/Makefile [hunk 14]
```diff
@@ -772,17 +772,19 @@ min_path_size=200
 	mlr --tsvlite filter '$$n >= $n' $< >$@
 
 # Separate barcodes into molecules.
-mol_strategy=1
+mol_strategy=bc
 %.mol.tsv: %.tsv
 	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=$(mol_strategy) $< >$@
 
 # Separate barcodes into molecules using K3-clique community detection.
+mol_strategy2=bc+k3
```
Comment (sjackman) [inline data/Makefile]:
```suggestion
```

File: data/Makefile [hunk 15]
```diff
@@ -772,17 +772,19 @@ min_path_size=200
...
 %.k3.mol.tsv: %.tsv
-	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=2 $< >$@
+	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=$(mol_strategy2) $< >$@
```
Comment (sjackman) [inline data/Makefile]:
```suggestion
	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=bc+k3 $< >$@
```

File: data/Makefile [hunk 16]
```diff
@@ -772,17 +772,19 @@ min_path_size=200
...
 
 # Separate barcodes into molecules using Louvain community detection.
+mol_strategy3=bc+louvain
```
Comment (sjackman) [inline data/Makefile]:
```suggestion
```

File: data/Makefile [hunk 17]
```diff
@@ -772,17 +772,19 @@ min_path_size=200
...
 %.louvain.mol.tsv: %.tsv
-	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=3 $< >$@
+	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=$(mol_strategy3) $< >$@
```
Comment (sjackman) [inline data/Makefile]:
```suggestion
	$(time) $(python) $(bin)/physlr molecules -V$V -t$t --separation-strategy=bc+louvain $< >$@
```

**Pull request merged**